File: NTerraform/Resources/AzurermVirtualNetwork.cs
using System.Collections.Generic;

namespace NTerraform.Resources
{
    public sealed class azurerm_virtual_network : NTerraform.resource
    {
        public sealed class subnet
        {
            public subnet(string @addressPrefix,
                          string @name,
                          string @securityGroup = null)
            {
                @AddressPrefix = @addressPrefix;
                @Name = @name;
                @SecurityGroup = @securityGroup;
            }

            public string @AddressPrefix { get; }
            public string @Name { get; }
            public string @SecurityGroup { get; }
        }

        public azurerm_virtual_network(string[] @addressSpace,
                                       string @location,
                                       string @name,
                                       string @resourceGroupName,
                                       string[] @dnsServers = null,
                                       subnet[] @subnet = null)
        {
            @AddressSpace = @addressSpace;
            @Location = @location;
            @Name = @name;
            @ResourceGroupName = @resourceGroupName;
            @DnsServers = @dnsServers;
            @Subnet = @subnet;
        }

        public string[] @AddressSpace { get; }
        public string @Location { get; }
        public string @Name { get; }
        public string @ResourceGroupName { get; }
        public string[] @DnsServers { get; }
        public subnet[] @Subnet { get; }
        public Dictionary<string,string> @Tags { get; }
    }

}

File: SandboxFS/AzurermVirtualNetwork.fs
module AzurermVirtualNetwork

type Subnet(addressPrefix : string,
            name : string,
            ?securityGroup : string) =
    member this.AddressPrefix = addressPrefix
    member this.Name = name
    member this.SecurityGroup = securityGroup
    
type AzurermVirtualNetwork(addressSpace: string list, 
                           location : string, 
                           name : string, 
                           resourceGroupName : string,
                           ?dnsServers : string list,
                           ?tags : Map<string, string>,
                           ?subnets : Subnet list) =
    inherit NTerraformCore.Resource()

    member this.AddressSpace = addressSpace
    member this.Location = location
    member this.Name = name
    member this.ResourceGroupName = resourceGroupName
    member this.DnsServers = dnsServers
    member this.Tags = tags
    member this.Subnets = subnets
